### PR TITLE
Nullary RNGOp

### LINKED ISSUE #1891: randlike gives wrong output dtype
### 🐛 Describe the bug

```C++
TEST_F(NVFuserTest, TMP4) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  TensorView* tv0 = makeSymbolicTensor(1);
  auto tv1 = randlike(tv0);
  std::cout << tv0->getDataType().value() << ", " << tv1->getDataType().value() << std::endl;
  TORCH_CHECK(tv0->getDataType() == tv1->getDataType());
}
```
I get `float, double`

### Versions

devel

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 1]
```diff
@@ -720,13 +727,29 @@ class ValReplacementMutator : private OptOutMutator {
     // would be a tensorview that doesn't get updated extents. Therefore, first
     // grab all leaves towards outputs and grab stmts from there.
     auto stmts = StmtSort::getStmts(fusion, allLeafOuts(fusion), true);
-    for (auto stmt : stmts) {
+
+    // Some fusions, such as standalone randlike, can have disconnected DAG, so
+    // we need some mechanism to make sure our replacement set is as complete as
+    // possible
+    // TODO: I think we need a more general mechanism to support disconnected
+    // DAG
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp:735]: WDYM for disconnected DAG? If we had two DAGs that don't connect in the same fusion, and we had outputs in each, wouldn't we grab all the values produced by the RNG expr? Confused why this code is needed, I would think StmtSort would grab all the nodes needed.

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 2]
```diff
@@ -720,13 +727,29 @@ class ValReplacementMutator : private OptOutMutator {
...
+    std::vector<Val*> more;
+    for (auto v : fusion->inputs()) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp:737]: Inputs aren't already in the `stmts` from `StmtSort`?

File: torch/csrc/jit/codegen/cuda/kernel.cpp [hunk 3]
```diff
@@ -80,11 +80,11 @@ class KernelIrScanner : private IrVisitor {
     }
   }
 
-  void handle(UnaryOp* unary_op) final {
-    if (unary_op->getUnaryOpType() == UnaryOpType::RandLike) {
-      summary_.max_rng_offsets =
-          std::max<int>(summary_.max_rng_offsets, unary_op->getRNGOffset());
-    }
+  void handle(UnaryOp* unary_op) final {}
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel.cpp]: Can you just remove this now?

File: torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp [hunk 4]
```diff
@@ -14,8 +14,9 @@ void ConcretizedBroadcastDomains::build(Fusion* fusion) {
   exact_map_ = std::make_unique<ExactRootDomainMap>(fusion);
 
   // Initialize the origin map with input broadcast domains
+  auto inputs = fusion->inputsAndCreated();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:17]: Are there other cases like this? Do we need to audit all `fusion->inputs()` calls?

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 5]
```diff
@@ -1005,6 +1019,11 @@ void ComputeAtRootDomainMapBuilder::mapAllPendingMappings(
   }
 }
 
+void ComputeAtRootDomainMapBuilder::handle(RNGOp* rop) {
+  const TensorDomain* td = rop->output(0)->as<TensorView>()->domain();
+  mapAllPendingMappings(td);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: Why not just call `handle(TensorView* tv)` on `rop->output(0)->as<TensorView>()`?

**Review by csarofeen — APPROVED** [review]
I think this should be fine to take in though I'm still a bit confused around the `StmtSort` comment, and generally concerned about everywhere we call to `fusion->inputs()`.

File: torch/csrc/jit/codegen/cuda/kernel.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/kernel.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/root_domain_map.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.cpp:737]: They are not. For example
```
input  T0
output T1 = randlike(T0)
```
Then `T0` and `T1` will be disconnected. For such case
```c++
StmtSort::getStmts(fusion, allLeafOuts(fusion), true);
```
will not include `T0`

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_utils.cpp:735]: See https://github.com/csarofeen/pytorch/pull/1892#discussion_r952206918

File: torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp:17]: I skimmed through our codebase and feel that many of them are hard to tell if it needs to handle this or not. So I only changed the one that has a failing test.

For examples:

In the reduction and normalization scheduler, I see:
```C++
  auto tv_inps = ir_utils::filterByType<TensorView>(fusion->inputs());
  TORCH_INTERNAL_ASSERT(
      !tv_inps.empty(),
      "Tried to schedule a fusion with no tensor inputs, currently not supported.");
```
we should change this, but I don't know how much work is needed.

**Pull request merged**